### PR TITLE
batches: Mount paths into a container

### PR DESCRIPTION
**Pull request by Piszmog**
Closes [#31790](https://github.com/sourcegraph/sourcegraph/issues/31790).

Add the ability to mount a path on the local machine into a container during a `step`. 

A `path` can point to a file (`./foo.sh` or  `/bar/foo.sh`) or a directory (`./bar` or `/foo/bar/`). The `path` can be an absolute path or a relative path (for example `./foo.sh`). Regardless if the path is absolute or relative, the path must be within the same directory as the batch spec that is being ran (the batch spec directory is considered the "working directory").

Validation/handling of the `mount` field in a `step` is handled in the `src-cli` versus `batcheslib` in `sourcegraph/sourcegraph`. The reason for this was the avoid having conditional logic to figure out if `parseBatchSpec` is being called on a local machine vs behind an API. This also contains the file system logic within the context of `src-cli` which is dealing with the file system.

This PR also includes some general fixes I saw while in the code (for example, package and variable naming collision).

### Test plan

Go Unit Tests and manual testing.

#### Examples

Below are examples of mounting a path that contains a script (python or Go binary in this case) that is ran in the `run` field of a `step`.

##### Python

Example of running a python script that is mounted into the container and ran in a `step`.

###### Batch Spec

```yaml
name: python-script-test
description: Use a python script
on:
  - repositoriesMatchingQuery: file:README.md
steps:
  - run: python /tmp/updater.py
    container: python:latest
    mount:
      - path: ./updater.py
        mountpoint: /tmp/updater.py
changesetTemplate:
  title: Hello from Python
  body: Updated using a Python Script
  branch: python-script-test
  commit:
    message: Append Hello World to all README.md files
```

###### Script

```python
#!/usr/bin/env python3
import os.path


def main():
    exists = os.path.exists('README.md')
    if exists:
        with open('README.md', 'a') as f:
            f.write('\nHello from python')


if __name__ == "__main__":
    main()

```

###### Changesets

Screen Shot 2022-06-07 at 14 43 47

##### Binary

Example of running a binary (written in Go) that is mounted into the container and ran in a `step`.

###### Batch Spec

```yaml
name: binary-test
description: Use a binary
on:
  - repositoriesMatchingQuery: file:README.md
steps:
  - run: /tmp/updater
    container: alpine:latest
    mount:
      - path: ./updater # the linux binary
        mountpoint: /tmp/updater
changesetTemplate:
  title: Hello from a binary
  body: Updated using a binary
  branch: binary-test
  commit:
    message: Append Hello World to all README.md files
```

###### Binary

Binary was built from Go

```go
package main

import "os"

func main() {
	fileInfo, err := os.Stat("README.md")
	if err != nil {
		panic(err)
	}
	if fileInfo.IsDir() {
		panic("file is a directory")
	}

	f, err := os.OpenFile("README.md", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	if _, err = f.WriteString("\nhello from go"); err != nil {
		panic(err)
	}
}
```

Built with the following command,

`env GOOS=linux GOGOARCH=amd64 go build -ldflags="-s -w"`

###### Changesets

Screen Shot 2022-06-07 at 14 44 50

##### Directory

The above examples have the `path` pointing to a specific file. Directories can also be mounted in the container. This example reuses the above Python Example but mounts a directory that contains the script.

###### Batch Spec

```yaml
name: python-script-test
description: Use a python script
on:
  - repositoriesMatchingQuery: file:README.md
steps:
  - run: python /tmp/scripts/updater.py
    container: python:latest
    mount:
      - path: ./scripts
        mountpoint: /tmp/scripts
changesetTemplate:
  title: Hello from Python
  body: Updated using a Python Script
  branch: python-script-test
  commit:
    message: Append Hello World to all README.md files
```

###### Changesets

Screen Shot 2022-06-07 at 14 42 45

### CODE REVIEW

**Comment by courier-new** [comment]
Thank you for such a thorough write-up of your testing examples!! 😄

File: internal/batches/service/service.go
Comment (courier-new) [inline internal/batches/service/service.go:385]:
```suggestion
		return nil, errors.Wrap(err, "handling mount")
```

File: internal/batches/service/service_test.go
Comment (courier-new) [inline internal/batches/service/service_test.go:750]: Could it be worth adding a test case with multiple steps/multiple mounted files as well?

**Review by courier-new — APPROVED** [review]
I defer to others on the team for a heavier review of the docker-specific interactions, but the mounted file handling and other changes look great to me. 🙂

File: internal/batches/executor/run_steps.go
Comment (eseliger) [inline internal/batches/executor/run_steps.go:309]: Can we add a flag to `executionOpts` like `allowFileMounts` that we wire through the executor and coordinator layer, so that we can explicitly disable it in `batch_exec.go` (the entrypoint for SSBC?) Feels like a good safety measure, in case someone can still sneak a batch spec into the server that contains mounts (I think we currently don't reject them in the SSBC `createBatchSpecFromRaw` API, just in the editor UI)

File: internal/batches/service/service_test.go
Comment (eseliger) [inline internal/batches/service/service_test.go:969]: great test suite!

File: internal/batches/service/service.go
Comment (eseliger) [inline internal/batches/service/service.go:390]: This sounds like a more precise description
```suggestion
func validateMountPaths(safeDirectory string, spec *batcheslib.BatchSpec) error {
```

File: cmd/src/batch_common.go
Comment (eseliger) [inline cmd/src/batch_common.go:9]: There's a file called `batch_remote` in this directory that uploads the spec file for SSBC execution, we should add a gate there that the `validateMountPaths` method rejects if there is a mount entry present. The API should ideally reject this anyways (to be implemented), but a nice actionable error message can't hurt, I think.

File: cmd/src/batch_common.go
Comment (LawnGnome) [inline cmd/src/batch_common.go:486]: For clarity, I probably would have gone with:

```suggestion
	dir := filepath.Dir(p)
```

Unless there's a behavioural quirk in one of `Dir` or `Split` that I've forgotten. (Which is entirely possible!)

File: cmd/src/batch_common.go
Comment (eseliger) [inline cmd/src/batch_common.go:482]: `file` can also be an empty string or `-`, which indicate "read from stdin". Should we in that case fall back to the working directory? Or error? Regardless, one of the two we need to do here.

**Review by eseliger — COMMENTED** [review]
Great PR description and testing! I left a few comments inline

File: internal/batches/executor/coordinator.go
Comment (LawnGnome) [inline internal/batches/executor/coordinator.go:26]: :x: no longer marks the spot.

File: internal/batches/executor/executor.go
Comment (LawnGnome) [inline internal/batches/executor/executor.go:151]: Good catch — I often shadow a package name inadvertently and then wonder why things break later.

File: internal/batches/service/service.go
Comment (LawnGnome) [inline internal/batches/service/service.go:404]: I'm wondering if we'll have edge cases here around symlinks — for example, if the user has their spec in `/Users/foo/bar`, but there's a symlink `/Users/foo/quux` to it and they use the latter in an absolute path provided to `mount`.

The quick and easy solution here might be to call [`filepath.EvalSymlinks`](https://pkg.go.dev/path/filepath#EvalSymlinks) on `p` and `dir` before performing this check. (I guess we only really need to do `dir` once.)

File: cmd/src/batch_common.go
Comment (LawnGnome) [inline cmd/src/batch_common.go:482]: We'll need to special case `file == "-"` and `file == ""`, per the code in `batchOpenFileFlag` to handle reading from stdin. (In that case, `dir` is presumably going to be the current working directory.)

File: internal/batches/executor/run_steps.go
Comment (LawnGnome) [inline internal/batches/executor/run_steps.go:308]: Edge case: what happens if `mount.Path` or `mount.Mountpoint` include a literal comma?

**Review by LawnGnome — APPROVED** [review]
This looks really good! I have a few questions, but nothing fundamental to how this is implemented.

File: internal/batches/executor/run_steps.go
Comment (eseliger) [reply internal/batches/executor/run_steps.go:308]: That sounds scary 😬 I think the same question applies to the `files` property.

File: internal/batches/service/service.go
Comment (eseliger) [reply internal/batches/service/service.go:404]: Hmm, don't you think that could actually be intentional? 

Like I put a symlink down to use it in a different WD without copying it?

Or do I misunderstand what the problem is here?

Comment (LawnGnome) [reply internal/batches/service/service.go:404]: I guess my question is more: should we consider `/Users/foo/bar` and `/Users/foo/quux` to be the same for the purposes of this check, if they point to the same directory?

Comment (eseliger) [reply internal/batches/service/service.go:404]: I, having thought about this at least an entire 30 seconds would say we should not. 

If you symlink `/Users/foo/spec.yml` into `/Users/erik/workspace/spec.yml` and use that one, then the intentional workspace location IMO would be `/Users/erik/workspace` and not `/Users/foo`. I only symlinked the spec file and not it's surroundings. If I symlinked the entire directory `/Users/foo` to `/Users/erik/workspace`, then I should just be using relative paths or paths to the aliased name. 
At least to my 12:30am brain that sounds logical. Please challenge me 😬

File: internal/batches/service/service.go
Comment (Piszmog) [reply internal/batches/service/service.go:390]: Initially I went with `validate` in the function name to match the other functionality going on. But this is a bit more than "validate" It is also updating the batch spec (changing the mount path to the absolute path). So I felt `handle` was a bit closer to the functionality.

But thinking more about it, I think `validate` is "correct" since that is like 99% of the functionality

File: internal/batches/service/service.go
Comment (Piszmog) [reply internal/batches/service/service.go:414]: I don't see any issues with caching, but I am not super familiar with the cache yet. 

Actually, I was having issues with the cache where I would only make a change to the script file and would rerun the batch spec and nothing would change because of the cache (for example changing the text to append to the `README.md` from `Hello from python` to `Hello from python1`).

I was getting slightly annoyed with this issue, and I am sure another user will too. Is this something we want to deal with in this Issue or in a follow up Issue?

File: internal/batches/executor/run_steps.go
Comment (Piszmog) [reply internal/batches/executor/run_steps.go:308]: If a comma is provided, Docker gets angry because it thinks the value after the comma is a field and it is not in the right format.

```text
invalid argument "type=bind,source=/Users/randell/Documents/development/batch-spec-playground/some,bad/foo.py,target=/tmp/some/foo.py,ro" for "--mount" flag: invalid field 'bad/foo.py' must be a key=value pair

invalid argument "type=bind,source=/Users/randell/Documents/development/batch-spec-playground/foo.py,target=/tmp/some,bad/foo.py,ro" for "--mount" flag: invalid field 'bad/foo.py' must be a key=value pair
```

Regardless, I'll add a check for this character and any others.

File: internal/batches/service/service.go
Comment (Piszmog) [reply internal/batches/service/service.go:414]: I just saw the flag `-clear-cache`. So I guess, it is really not a problem? Maybe we should have some docs around the mount and cache and communicate that the `-clear-cache` flag is available when updating the script[s]?

File: internal/batches/executor/run_steps.go
Comment (Piszmog) [reply internal/batches/executor/run_steps.go:309]: Yea, we are not rejecting `mount` in the API layer at all.

File: internal/batches/service/service.go
Comment (eseliger) [inline internal/batches/service/service.go:417]: Can we move the static checks in here into the lib parse method, so SSBC can benefit from the same validation? Only works for the ones that don't need to access the FS, of course.

File: internal/batches/executor/executor.go
Comment (eseliger) [inline internal/batches/executor/executor.go:67]: I think this one you also want to restrict when mounts are used. Otherwise, step-wise caching wouldn't trigger on changed files either.

**Review by eseliger — APPROVED** [review]
Great, very clean :)

File: internal/batches/service/service.go
Comment (Piszmog) [reply internal/batches/service/service.go]: I'll need to wait for the `sourcegraph` PR to be merged so I can update `go.mod` to the latest version

**Pull request merged**
